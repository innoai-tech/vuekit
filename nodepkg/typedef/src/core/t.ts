import {TypeRef} from "./TypeRef.ts";
import {TypeAny} from "./TypeAny.ts";
import {DefaultedType, OptionalType, Type, TypeNever, TypeWrapper} from "./Type.ts";
import {TypeNull} from "./TypeNull.ts";
import {TypeString} from "./TypeString.ts";
import {TypeNumber} from "./TypeNumber.ts";
import {TypeInteger} from "./TypeInteger.ts";
import {TypeBoolean} from "./TypeBoolean.ts";
import {TypeBinary} from "./TypeBinary.ts";
import {TypeEnum} from "./TypeEnum.ts";
import {TypeObject} from "./TypeObject.ts";
import {TypeRecord} from "./TypeRecord.ts";
import {TypeArray} from "./TypeArray.ts";
import {TypeTuple} from "./TypeTuple.ts";
import {TypeIntersection} from "./TypeIntersection.ts";
import {TypeUnion} from "./TypeUnion.ts";

export const ref = TypeRef.create;
export const any = TypeAny.create;
export const never = TypeNever.create;
export const nil = TypeNull.create;

export const string = TypeString.create;

export function pattern(pattern: RegExp, msg?: string) {
    return <T extends string, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (pattern.test(value)) {
                return true
            }
            return msg ?? `Expected a ${type.type} matching \`/${pattern.source}/\` but received "${value}"`
        }, {
            pattern,
        })
    };
}

export const number = TypeNumber.create;
export const integer = TypeInteger.create;

export function minimum<M extends number>(min: M, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value >= min) {
                return true
            }
            return msg ?? `Expected value great than or equal ${min}, but received "${value}"`
        }, {
            minimum: min,
        })
    };
}

export function exclusiveMinimum<M extends number>(min: M, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value > min) {
                return true
            }
            return (msg ?? `Expected value great than ${min}, but received "${value}"`)
        }, {
            exclusiveMinimum: min,
        })
    };
}


export function maximum(max: number, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value <= max) {
                return true
            }
            return (msg ?? `Expected value less than or equal ${max}, but received "${value}"`)
        }, {
            maximum: max,
        }) as any
    };
}


export function exclusiveMaximum(max: number, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value < max) {
                return true
            }
            return msg ?? `Expected value less than or equal ${max}, but received "${value}"`
        }, {
            exclusiveMaximum: max,
        }) as any;
    };
}


export function multipleOf(multipleOf: number, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value % multipleOf == 0) {
                return true
            }
            return msg ?? `Expected value multiple of ${multipleOf}, but received "${value}"`
        }, {
            multipleOf: multipleOf,
        });
    };
}

export const boolean = TypeBoolean.create;
export const binary = TypeBinary.create;

export const enums = TypeEnum.create;
export const nativeEnum = TypeEnum.nativeEnum;
export const literal = TypeEnum.literal;

export const object = TypeObject.create;
export const record = TypeRecord.create;

export const tuple = TypeTuple.create;
export const array = TypeArray.create;

export function minItems(minItems: number, msg?: string) {
    return <T extends any[], S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (Array.isArray(value) && value.length >= minItems) {
                return true
            }
            return msg ?? `Expected array value at least ${minItems}, but received "${value?.length}"`
        }, {
            minItems: minItems,
        })
    };
}

export function maxItems(maxItems: number, msg?: string) {
    return <T extends any[], S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (Array.isArray(value) && value.length <= maxItems) {
                return true
            }
            return msg ?? `Expected array value  ${maxItems}, but received "${value?.length}"`
        }, {
            maxItems: maxItems,
        });
    };
}

export const intersection = TypeIntersection.create;
export const union = TypeUnion.create;
export const discriminatorMapping = TypeUnion.discriminatorMapping;

export const custom = Type.define;
export const refine = TypeWrapper.refine;

export function defaults<T>(v: T) {
    return <U extends Type<T, any>>(type: U): U => {
        return DefaultedType.create(type, v) as unknown as U;
    };
}

export function optional() {
    return <T, S>(type: Type<T, S>): Type<T | undefined, S> => {
        return OptionalType.create(type) as unknown as any;
    };
}

export function annotate<M extends Record<string, any>>(meta: M) {
    return <T, S>(type: Type<T, S>): Type<T, S & M> => {
        return TypeWrapper.of(type, {$meta: meta}) as unknown as Type<T, S & M>;
    };
}
