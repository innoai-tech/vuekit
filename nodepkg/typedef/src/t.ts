export * from "./types";
export * from "./modifers";

