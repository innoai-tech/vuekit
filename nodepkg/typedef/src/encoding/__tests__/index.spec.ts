import { describe, test, expect } from "vitest";
import * as t from "../../t";
import {
  TypeScriptEncoder,
  JSONSchemaDecoder,
  JSONSchemaEncoder,
  refName,
  TypedefEncoder
} from "../";
import { get } from "@innoai-tech/lodash";

describe("Encoding", () => {
  enum InputType {
    text = "text",
    number = "number",
    select = "select",
  }

  const schemaStrOrInt = t.union(t.string(), t.integer());

  const schemaTaggedUnion = t.discriminatorMapping("type", {
    [InputType.text]: t.object(),
    [InputType.select]: t.ref("WithOptions", () =>
      t.object({
        options: t.array(
          t.object({
            label: t.string(),
            value: t.string()
          })
        )
      })
    )
  });

  const schema = t.intersection(
    t.object({
      strOrInt: t
        .ref("StrOrInt", () => schemaStrOrInt)
        .annotate({
          description: "StrOrInt"
        }),
      placement: t.enums(["leading", "trailing"] as const),
      inputType: t.ref("InputType", () => t.nativeEnum(InputType)).optional(),
      keyValues: t.record(t.string(), t.any()).optional(),
      array: t.array(t.boolean()),
      point: t.tuple([t.number(), t.number()])
    }),
    schemaTaggedUnion
  );

  test("JSONSchema decode", () => {
    const jsonSchema = JSONSchemaEncoder.encode(schema);

    const schema2 = JSONSchemaDecoder.decode(jsonSchema, (ref) => {
      return [get(jsonSchema, ref.split("#/")[1]!.split("/")), refName(ref)];
    });

    const jsonSchema2 = JSONSchemaEncoder.encode(schema2);
    expect(jsonSchema2).toEqual(jsonSchema);
  });

  test("Typedef encode", () => {
    const typeDefCode = TypedefEncoder.encode(t.ref("Type", () => schema));
    expect(typeDefCode).toMatchSnapshot();
  });

  test("TypeScript encode", () => {
    const tsCode = TypeScriptEncoder.encode(t.ref("Type", () => schema));
    expect(tsCode).toMatchSnapshot();
  });

  test("JSONSchema encode", () => {
    const jsonSchema = JSONSchemaEncoder.encode(schema);
    console.log(JSON.stringify(jsonSchema, null, 2));
    expect(jsonSchema).toMatchSnapshot();
  });
});
