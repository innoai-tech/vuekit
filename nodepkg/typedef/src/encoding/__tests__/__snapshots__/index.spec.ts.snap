// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Encoding Typedef encode 1`] = `
"
      
export const TypeSchema = /*#__PURE__*/t.intersection(t.object({
  "strOrInt": t.ref<typeof StrOrIntSchema>("StrOrInt", () => StrOrIntSchema).use(t.annotate({ description: "StrOrInt" })),
  "placement": t.enums(["leading", "trailing"]),
  "inputType": t.ref<typeof InputTypeSchema>("InputType", () => InputTypeSchema).optional(),
  "keyValues": t.record(t.string(), t.any()).optional(),
  "array": t.array(t.boolean()),
  "point": t.tuple([t.number(), t.number()]),
}), t.discriminatorMapping("type", {
"text": t.object(),
"select": t.object({
  "options": t.array(t.object({
  "label": t.string(),
  "value": t.string(),
})),
})
          }))
      
export const StrOrIntSchema = /*#__PURE__*/t.union(t.string(), t.integer())
      
export const InputTypeSchema = /*#__PURE__*/t.nativeEnum(InputType)"
`;

exports[`Encoding TypeScript encode 1`] = `
"
      
export type Type = ({
  "strOrInt": /* @type:union */ StrOrInt,
  "placement": "leading" | "trailing",
  "inputType"?: /* @type:enums */ InputType,
  "keyValues"?: { [k: string]: any },
  "array": Array<boolean>,
  "point": [number, number],
} & ({
  "type": "text",
} | {
  "type": "select",
  "options": Array<{
  "label": string,
  "value": string,
}>,
}))
      
export type StrOrInt = (string | number)
      
export enum InputType {
text = "text",
number = "number",
select = "select"         
}
      
export const displayInputType = (v: InputType) => {
  return ({
"text": "文本",
"number": "数字",
"select": "选项"   
  })[v] ?? v      
}"
`;

exports[`Encoding JSONSchema encode 1`] = `
{
  "allOf": [
    {
      "additionalProperties": false,
      "properties": {
        "array": {
          "items": {
            "type": "boolean",
          },
          "type": "array",
        },
        "inputType": {
          "$ref": "#/definitions/InputType",
        },
        "keyValues": {
          "additionalProperties": {},
          "propertyNames": {
            "type": "string",
          },
          "type": "object",
        },
        "placement": {
          "enum": [
            "leading",
            "trailing",
          ],
        },
        "point": {
          "items": [
            {
              "type": "number",
            },
            {
              "type": "number",
            },
          ],
          "type": "array",
        },
        "strOrInt": {
          "$ref": "#/definitions/StrOrInt",
          "description": "StrOrInt",
        },
      },
      "required": [
        "strOrInt",
        "placement",
        "array",
        "point",
      ],
      "type": "object",
    },
    {
      "discriminator": {
        "propertyName": "type",
      },
      "oneOf": [
        {
          "additionalProperties": false,
          "properties": {
            "type": {
              "enum": [
                "text",
              ],
            },
          },
          "required": [
            "type",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "properties": {
            "options": {
              "items": {
                "additionalProperties": false,
                "properties": {
                  "label": {
                    "type": "string",
                  },
                  "value": {
                    "type": "string",
                  },
                },
                "required": [
                  "label",
                  "value",
                ],
                "type": "object",
              },
              "type": "array",
            },
            "type": {
              "enum": [
                "select",
              ],
            },
          },
          "required": [
            "type",
            "options",
          ],
          "type": "object",
        },
      ],
    },
  ],
  "definitions": {
    "InputType": {
      "enum": [
        "text",
        "number",
        "select",
      ],
    },
    "StrOrInt": {
      "oneOf": [
        {
          "type": "string",
        },
        {
          "type": "integer",
        },
      ],
    },
  },
}
`;
